xmrstak/backend/amd/amd_gpu/gpu.cpp

```diff
@@ -448,39 +451,60 @@ std::vector<GpuContext> getAMDDevices(int index)
 
 	uint32_t numPlatforms = getNumPlatforms();
 
-
-	platforms = (cl_platform_id *) malloc(sizeof(cl_platform_id) * numPlatforms);
-	clStatus = clGetPlatformIDs(numPlatforms, platforms, NULL);
-
-	clStatus = clGetDeviceIDs( platforms[index], CL_DEVICE_TYPE_GPU, 0, NULL, &num_devices);
-	device_list = (cl_device_id *) malloc(sizeof(cl_device_id)*num_devices);
-	clStatus = clGetDeviceIDs( platforms[index], CL_DEVICE_TYPE_GPU, num_devices, device_list, NULL);
-	for (int k = 0; k < num_devices; k++) {
-		cl_int clError;
-		std::vector<char> devVendorVec(1024);
-		clError = clGetDeviceInfo(device_list[k], CL_DEVICE_VENDOR, devVendorVec.size(), devVendorVec.data(), NULL);
-		std::string devVendor(devVendorVec.data());
-		if( devVendor.find("Advanced Micro Devices") != std::string::npos)
+	if(numPlatforms)
```

Comment:
Accepting right now, but this will need to be reformatted.
Good way to do it is:
```
std::vector<GpuContext> getAMDDevices(int index) -> bool getAMDDevices(int index, std::vector<GpuContext>& ctxVec)

if(!numPlatforms)
    return false;
```

Reply:
You are right I wrote it on this way to avoid that I need to call `free` to the created heap buffer in the case of an error/warning.
The full files needs a cleanup. There is a lot of dead code in and many code duplication's which should be removed to increase the maintenance.

Reply:
This is actually a good usecase for a dynamic structure, you can avoid it by doing `std::vector<uint_8t> buffer; buffer.resize(len); buffer.data()`